First draft of user-guide for compositions (#8)
cloud-native-compositions/compositions: docs/overview.md

# Overview

## Goals

> Help infrastructure platform teams build self-service developer platforms that work consistently across clouds. We are doing this by creating a K8s-native way of orchestrating cloud services and kubernetes resources.

Reviewer: Agree this is "the goal", but wondering if we want to be a bit more fine-grained than this, with something like a bulleted list?

Not 100% sure what our organizational plan is yet (probably we don't have one), but I'd expect somewhere to see in the list of goals some points that cannot be satisfied by existing tools.

Maybe:
```suggestion
- Help infrastructure platform teams build self-service developer platforms that work consistently across clouds.
- Provide a Kubernetes-native way of orchestrating cloud services and Kubernetes resources.
```

It also may be worth stealing/cribbing from the goals in the google doc going around, which are:

```
Empower platform teams to build K8s-native internal developer platforms that: 

- Abstract and resolve complex interdependencies between resources, enabling end user self-service for platform teams and simplified resource management for small dev teams
- Provide visibility into how an application is interacting with cloud services
- Are easier and cheaper to build and maintain than existing solutions
- Are compatible with any K8s operator, and existing packaging and templating tools
- Leverage existing extensibility types (CRDs) as the API for interacting with cloud resources

```

Author: Love this. Thanks. Was thinking of it. We should definitely revise this doc as we agree on the goals. For now crib list is a good starting point. Thanks for the suggestion.

> ## Customer Context

Reviewer: This might just be a me thing, but I (generally) prefer the term user over customer.

Author: done

> we have found that our customers often use  ACK/ASO/KCC as an input into a custom-built internal developer platform (IDP). In nearly all cases, platform teams are maintaining a custom orchestration layer between the developer interface and the cloud resources. The orchestration layer is typically responsible for: 

Reviewer: ```suggestion
We have found that our customers often use ACK/ASO/KCC as an input into a custom-built internal developer platform (IDP). In nearly all cases, platform teams are maintaining a custom orchestration layer between the developer interface and the cloud resources. The orchestration layer is typically responsible for: 
```

Also, ACK/ASO/KCC should be spelled out the first time as not everybody may know what they are, and also should _probably_ link to the corresponding GH repos of each.

Author: Made the changes.

* RBAC
* Resource creation ordering
* Continuous reconciliation
* Hydration of templates with developer inputs
* Security checks
* Automated and manual approval workflows
* Abstracting platform complexity for developer ease-of-use 

> There is no consistent K8s-native set of tooling used to build this orchestration layer. The most common solution is to write a layered set of Helm charts, with the top layer exposing only the parameters available to developers. Existing solutions tend to be challenging to maintain and require additional investment in hydration pipelines and custom RBAC models. Customers have been asking us to help address the challenges of orchestrating across different cloud services and teams / roles. The solution must be flexible/extensible so that existing solutions can be “plugged in”, but relatively simple to use out-of-the box in a greenfield scenario. Additionally, our customers have been asking for a consistent approach to cloud resource orchestration across cloud service vendors.

Reviewer: " Existing solutions tend to be challenging to maintain and require additional investment in hydration pipelines and custom RBAC models."

I don't know if this is the place, but somewhere it would be good to say _why_ they're challenging to maintain. What specific maintenance problems are we solving?

Author: @nicslatts - Do you have thoughts on the this comment ?

Reviewer 2: probably it's more accurate to say they are more challenging to scale and troubleshoot. Troubleshooting a values.helm file can be a nightmare, as well as detecting whether values are even being honored at all, and other solutions like Terraform are harder to scale because they are originally intended for human usage rather than automated pipelines.

## Product Description
Compositions provides platform administrators with a framework for building  secure, qualified and reusable APIs for application teams to consume cloud resources in a self-service manner. Compositions can be used as the orchestration layer below an IDP, or they can be used directly by developers either through the K8s CLI or through a GitOps tool such as Config Sync or Argo CD. 

### How Compositions Work

> Compositions aims to provide a K8s-based solution for the problem of self-service infrastructure management. It is a set of Kubernetes APIs (CRDs) and controllers that allow any K8s resources, including KCC/ASO/ACK resources, to be logically grouped together and consumed as a single resource (a “composition”). By building compositions as K8s-native resources, compositions can use K8s features such as RBAC and continuous reconciliation. 

Reviewer: Is there also an aspect of "and be monitored and administered the same way as other Kubernetes resources, which allows them to plug into existing workflows more easily"

Author: For sure there is. Incorporating the suggestion.

Compositions will be built as OSS and will be available as a managed service through Config Controller. This private preview will use Config Controller to install all the components necessary to build compositions across Google Cloud, AWS, and Azure. 

Resources available to be managed through compositions by default are any resources available through KCC, ACK, and ASO, as well as all K8s-native constructs (such as namespaces and pods). KCC, ACK, and ASO provide mappings to Google Cloud, AWS, and Azure APIs respectively to Kubernetes - think of these as the building blocks for compositions. Compositions allow you to build your own abstractions that consume these building blocks. Compositions can also manage resources provided by other K8s operators, if you wish to bring your own operators. 

### Design Principles
Some of the design principles we aim to adhere to are:
* For life cycle management of cloud resources, delegate to lower level controllers for individual cloud resources.

> * Consistent experience across clouds.

Reviewer: "Consistent experience across clouds." almost seems _too_ cloud focused to me?

Maybe
```suggestion
* Consistent experience across any Kubernetes resource, with a focus on KCC/ACK/ASO
```

Or something like that?

Author: like your wording.

* Extensible architecture for meeting customer choice of configuration language, tooling.
* Well defined core for quick development and Day 1 usage.
* Automate, yet provide control and visibility.
* Compatibility with K8s ecosystem tooling, RBAC, gitops workflows, support shift-left patterns etc.

### Overall Architecture
The below diagram captures the overall architecture of compositions.

![image](architecture.png)

### Compositions Design
A composition is a YAML file that defines a set of resources and how they relate to each other. The key concepts and APIs that will be used in this private preview are described below.

#### Key Concepts:

> 1. **Composition** \- Defined by a platform admin, a composition is a YAML file that defines a set of resources and any logic (such as ordering, loops, cross-referencing) required to create a service that can be used by app development teams. A composition has:  

Reviewer: "a composition is a YAML file"

Is it a YAML file or a CRD? It's a CRD right?

Author: It is :)

Reviewer: Noticed you didn't change the language here though - do you want the focus to be on the YAML experience (or are you concerned about audience and usage of the somewhat nerdy CRD term?)

# Overview

## Goals

Empower platform teams to build K8s-native internal developer platforms that: 

- Abstract and resolve complex interdependencies between resources, enabling end user self-service for platform teams and simplified resource management for small dev teams
- Provide visibility into how an application is interacting with cloud services
- Are easier and cheaper to build and maintain than existing solutions
- Are compatible with any K8s operator, and existing packaging and templating tools
- Leverage existing extensibility types (CRDs) as the API for interacting with cloud resources
- Provide a Kubernetes-native way of orchestrating cloud services and Kubernetes resources.

## User Context

> We have found that our customers often use [Aws Controllers for Kubernetes(ACK)](https://aws.amazon.com/blogs/containers/aws-controllers-for-kubernetes-ack/), [Azure Service Operator(ASO)](https://azure.github.io/azure-service-operator/), [GCP Config Connector(KCC)](https://cloud.google.com/config-connector/docs/overview) as an input into a custom-built internal developer platform (IDP). In nearly all cases, platform teams are maintaining a custom orchestration layer between the developer interface and the cloud resources. The orchestration layer is typically responsible for: 

Reviewer: minor: 
```suggestion
We have found that our customers often use [Aws Controllers for Kubernetes(ACK)](https://aws.amazon.com/blogs/containers/aws-controllers-for-kubernetes-ack/), [Azure Service Operator(ASO)](https://azure.github.io/azure-service-operator/), and [GCP Config Connector(KCC)](https://cloud.google.com/config-connector/docs/overview) as an input into a custom-built internal developer platform (IDP). In nearly all cases, platform teams are maintaining a custom orchestration layer between the developer interface and the cloud resources. The orchestration layer is typically responsible for: 
```
